core/src/main/scala/scalaz/Cord.scala

```diff
@@ -1,163 +1,55 @@
-   * Time complexity: O(1)
-   */
-  def :+(x: => String): Cord = cord(self :+ x)
-
-  /**
-   * Prepends a `String` to the beginning of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def +:(x: String): Cord = cord(x +: self)
-
-  /**
-   * Prepends a `Char` to the beginning of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def -:(x: Char): Cord = cord(String.valueOf(x) +: self)
-
-  /**
-   * Appends a `Char` to the end of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def :-(x: => Char): Cord = cord(self :+ String.valueOf(x))
-
-  /**
-   * Removes the first character of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def tail: Cord = drop(1)
-
-  /**
-   * Removes the last character of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def init: Cord = take(self.measure - 1)
-
-  /**
-   * Removes the first `n` characters from the front of this `Cord`.
-   * Time complexity: O(min N (N - n))
-   */
-  def drop(n: Int): Cord = split(n)._2
-
-  /**
-   * Returns the first `n` characters at the front of this `Cord`.
-   * Time complexity: O(min N (N - n))
-   */
-  def take(n: Int): Cord = split(n)._1
-
-  /**
-   * Modifies each character in this `Cord` by the given function.
-   * Time complexity: O(N)
-   */
-  def map(f: Char => Char): Cord = cord(self map (_ map f))
-
-  def toList: List[Char] = toVector.toList
-  def toStream: Stream[Char] = toVector.toStream
-  def toVector: Vector[Char] = self.foldMap(_.toVector)
+sealed abstract class Cord {
   override def toString: String = {
-    val sb = new java.lang.StringBuilder(self.measure)
-    self.foreach { x =>
-      val _ = sb append x
-    }
+    val sb = new StringBuilder
+    appendTo(sb)
     sb.toString
   }
 
-  /** Transforms each character to a `Cord` according to the given function and concatenates them all into one `Cord`. */
-  def flatMap(f: Char => Cord): Cord = toVector.foldLeft(Cord())((as, a) => as ++ f(a))
+  // FIXME: stack safety. Try to at least make traversal on the right stack safe
+  // (split into two?)
+  private def appendTo(sb: StringBuilder): Unit = this match {
+    case Cord.Branch(a, b) =>
+      a.appendTo(sb): Unit
+      b.appendTo(sb): Unit
+    case Cord.Leaf(s) =>
+      val _ = sb.append(s)
+  }
 
-  /** Returns whether this `Cord` will expand to an empty string. */
-  def isEmpty: Boolean = !nonEmpty
+  // FIXME: used with the implicit... it should be removed
+  @deprecated("use the z / cord interpolators", "7.3")
+  def +:(o: Cord): Cord = Cord.Branch(o, this)
 
-  /** Returns whether this `Cord` will expand to a non-empty string. */
-  def nonEmpty: Boolean = self.iterator.exists(_.nonEmpty)
+  /** Strict evaluation variant of Monoid.append */
+  def ++(o: Cord): Cord = Cord.Branch(this, o)
 }
-
 object Cord {
-  private def cord[A](v: FingerTree[Int, String]): Cord = new Cord(v)
+  private final case class Branch(a: Cord, b: Cord) extends Cord
+  private final case class Leaf(s: String)            extends Cord
```

Comment:
I kinda wish we'd make these `public`; there aren't `Cord` invariants _per se_ (right?) and it may be useful to pattern-match on one (or at least, we don't know it won't).

Also, if we do that, it's probably a good idea to make smart-constructors so `Leaf("")` has type `Cord`.